python3Packages.tensorflow_2: 2.3.2 -> 2.4.0

ndl opened the pull request:
Also the following related changes:
* Removed Python 2 support because it's not supported by TF and related packages for a long time.
* Upgraded tensorboard and estimator packages to the required versions.
* Added extra plugins for tensorboard to support profiling.
* In the previous derivation versions, TF_SYSTEM_LIBS didn't have any effect because it was reset at repo fetching stage, so TF always used its own dependencies. Made TF_SYSTEM_LIBS actually work and fixed the errors caused by enabling it.
* Enabled tensorboard by default (but still keeping an option to disable it if needed).

###### Motivation for this change

To update and clean up TensorFlow package and dependencies.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

SuperSandro2000 commented on maintainers/maintainer-list.nix:
```diff
@@ -6385,6 +6385,12 @@
     githubId = 364510;
     name = "Tobias Geerinckx-Rice";
   };
+  ndl = {
```
Please put this into a separate commit with the message ``maintainers: add ndl``.

ndl replied:
Done.

SuperSandro2000 commented on pkgs/development/python-modules/flatbuffers/default.nix:
```diff
@@ -8,6 +8,10 @@ buildPythonPackage rec {
 
+  # flatbuffers needs VERSION environment variable for setting the correct
+  # version, otherwise it uses the current date.
+  VERSION = "${version}";
```
Can we move this into preBuild or similar with an export?

Mic92 replied:
I think it's fine to use a variable here because this also works with nix-shell than to get a build environment.

SuperSandro2000 commented on pkgs/development/python-modules/gviz-api/default.nix:
```diff
@@ -0,0 +1,26 @@
+  ];
+
+  meta = with lib; {
+    description = "This project is a helper python library for providers who want to implement a data source for visualizations built on the Google Visualization API.";
```
```suggestion
    description = "Python API for Google Visualization";
```

ndl replied:
Done.

SuperSandro2000 commented on pkgs/development/python-modules/tensorflow-tensorboard/default.nix:
```diff
@@ -16,19 +18,15 @@
   pname = "tensorflow-tensorboard";
-  version = "2.1.0";
+  version = "2.4.0";
   format = "wheel";
 
```
```suggestion
  format = "wheel";
  disabled = !isPy3k;

```

ndl replied:
Done.

SuperSandro2000 commented on pkgs/development/python-modules/tensorflow-tensorboard/default.nix:
```diff
@@ -8,6 +8,8 @@
 , futures
```
```suggestion
```

ndl replied:
Done.

SuperSandro2000 commented on pkgs/development/python-modules/tensorflow/default.nix:
```diff
@@ -1,24 +1,25 @@
 { stdenv, pkgs, bazel_3, buildBazelPackage, lib, fetchFromGitHub, fetchpatch, symlinkJoin
```
is fetchpatch now unused?

ndl replied:
Correct, not needed anymore - removed.

SuperSandro2000 commented on pkgs/development/python-modules/tensorflow/default.nix:
```diff
@@ -135,18 +135,17 @@ let
+      pybind11
       snappy
-      flatbuffers
+      pkgs.flatbuffers
```
```suggestion
      flatbuffers
```

Pleas inherit flatbuffers in top-level.

ndl replied:
Note that we need both "core" and "python" flatbuffers so we cannot just inherit "flatbuffers". Changed the names to "flatbuffers-core" and "flatbuffers-python", LMK if you prefer it done differently.

SuperSandro2000 commented on pkgs/development/python-modules/tensorflow/default.nix:
```diff
@@ -329,15 +332,12 @@ let
 
 in buildPythonPackage {
   inherit version pname;
-  disabled = isPy27;
```
Please put this back in.

ndl replied:
Done.

SuperSandro2000 commented on pkgs/top-level/python-packages.nix:
```diff
@@ -7386,9 +7392,7 @@ in {
     nccl = pkgs.nccl_cudatoolkit_11;
-    openssl = pkgs.openssl_1_1;
     inherit (pkgs.darwin.apple_sdk.frameworks) Foundation Security;
-    inherit (pkgs) flatbuffers;
```
Please revert this.

ndl replied:
Changed the names to avoid the conflict, see above.

Mic92 commented on pkgs/development/python-modules/tensorflow/default.nix:
```diff
@@ -103,27 +109,21 @@ let
 
     patches = [
-      # Fixes for NixOS jsoncpp
-      ./system-jsoncpp.patch
```
Does it still uses the system jsoncpp? If not please remove it from buildInputs.

Mic92 replied:
Please also remove the patch file

ndl replied:
It still should use system jsoncpp but it doesn't need the patch anymore. Removed the patch.

ndl committed:
maintainers: add ndl

ndl committed:
python3Packages.tensorflow_2: 2.3.2 -> 2.4.0

Also the following related changes:
* Removed Python 2 support because it's not supported by TF and related packages for a long time.
* Upgraded tensorboard and estimator packages to the required versions.
* Added extra plugins for tensorboard to support profiling.
* In the previous derivation versions, TF_SYSTEM_LIBS didn't have any effect because it was reset at repo fetching stage, so TF always used its own dependencies. Made TF_SYSTEM_LIBS actually work and fixed the errors caused by enabling it.
* Enabled tensorboard by default (but still keeping an option to disable it if needed).

Wulfsta commented:
These changes look good to me at a glance, but I cannot test because I have an AMD GPU. You might get asked to split this into commits for each package you modified or added.

Wulfsta commented on pkgs/development/python-modules/tensorflow/default.nix:
```diff
@@ -277,16 +281,15 @@ let
+    # Without this Bazel complaints about sandbox violations.
+    dontAddBazelOpts = true;
 
     fetchAttrs = {
```
Consider using `fetchConfigured`.

Wulfsta commented on pkgs/development/python-modules/tensorflow/default.nix:
```diff
@@ -277,16 +281,15 @@ let
 
     fetchAttrs = {
-      # So that checksums don't depend on these.
-      TF_SYSTEM_LIBS = null;
```
Why was this removed? I recall having issues trying to remove this in the past working on the ROCm TensorFlow derivation, but don't remember what those issues were. It might just be here to prevent changes to inputs into this derivation from changing these checksums.

ndl replied:
> Why was this removed?

The way TensorFlow switches between the system and bundled dependencies is that at fetch time it either downloads the dependency and creates a repository out of it, or (if the dependency is listed in `TF_SYSTEM_LIBS`) creates the "proxy" repository that points to the system library using the files bundled with the "main" TensorFlow source code, see https://github.com/tensorflow/tensorflow/tree/master/third_party/systemlibs.

This means that if `TF_SYSTEM_LIBS` is not set at fetch time, the build won't use the system libraries, so having it set in any section outside of "fetch" is basically a no-op.

> I recall having issues trying to remove this in the past working on the ROCm TensorFlow derivation

Given the above, in the derivation currently present in Nixpkgs, the build doesn't actually use any of the system libraries, so the build that uses system libraries was likely never tested before and enabling them uncovered several issues - like incorrect dependencies or incompatibilities between the libraries bundled with TF and provided in Nixpkgs (see the comments for "protobuf" and "re2" in the changelist). It also could very well be that there are ROCm-specific issues on top of this, though.

> It might just be here to prevent changes to inputs into this derivation from changing these checksums.

Given that these "proxy" repositories that point to system libraries shouldn't encode any version-specific information (normally they just encode the library name to link to, e.g. see https://github.com/tensorflow/tensorflow/blob/master/third_party/systemlibs/jsoncpp.BUILD) - I don't expect that changing the inputs versions will change the checksum of the dependencies?

Mic92 commented:
Result of `nixpkgs-review pr 108923` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python37Packages.rl-coach</li>
    <li>python38Packages.rl-coach</li>
  </ul>
</details>
<details>
  <summary>30 packages failed to build:</summary>
  <ul>
    <li>python37Packages.arviz</li>
    <li>python37Packages.baselines</li>
    <li>python37Packages.dm-sonnet</li>
    <li>python37Packages.edward</li>
    <li>python37Packages.graph_nets</li>
    <li>python37Packages.mask-rcnn</li>
    <li>python37Packages.optuna</li>
    <li>python37Packages.pymc3</li>
    <li>python37Packages.pytorch-metric-learning</li>
    <li>python37Packages.tensorflow (python37Packages.tensorflowWithoutCuda)</li>
    <li>python37Packages.tensorflow-bin</li>
    <li>python37Packages.tensorflow-probability</li>
    <li>python37Packages.tensorflowWithCuda</li>
    <li>python37Packages.tflearn</li>
    <li>python37Packages.torchgpipe</li>
    <li>python38Packages.arviz</li>
    <li>python38Packages.dm-sonnet</li>
    <li>python38Packages.graph_nets</li>
    <li>python38Packages.optuna</li>
    <li>python38Packages.pymc3</li>
    <li>python38Packages.tensorflow-probability</li>
    <li>python38Packages.tensorflowWithCuda</li>
    <li>python38Packages.torchgpipe</li>
    <li>python39Packages.dm-sonnet</li>
    <li>python39Packages.graph_nets</li>
    <li>python39Packages.pytorch-metric-learning</li>
    <li>python39Packages.pytorchWithCuda</li>
    <li>python39Packages.tensorflow-probability</li>
    <li>python39Packages.torchgpipe</li>
    <li>tts</li>
  </ul>
</details>
<details>
  <summary>62 packages built:</summary>
  <ul>
    <li>ocamlPackages.torch</li>
    <li>python37Packages.flatbuffers</li>
    <li>python37Packages.gviz-api</li>
    <li>python37Packages.ignite</li>
    <li>python37Packages.pyro-ppl</li>
    <li>python37Packages.pytorch (python37Packages.pytorchWithoutCuda)</li>
    <li>python37Packages.pytorch-lightning</li>
    <li>python37Packages.pytorchWithCuda</li>
    <li>python37Packages.pywick</li>
    <li>python37Packages.rising</li>
    <li>python37Packages.shap</li>
    <li>python37Packages.skorch</li>
    <li>python37Packages.slicer</li>
    <li>python37Packages.tensorboard-plugin-profile</li>
    <li>python37Packages.tensorboard-plugin-wit</li>
    <li>python37Packages.tensorboardx</li>
    <li>python37Packages.tensorflow-estimator</li>
    <li>python37Packages.tensorflow-tensorboard</li>
    <li>python37Packages.test-tube</li>
    <li>python37Packages.torchvision</li>
    <li>python38Packages.baselines</li>
    <li>python38Packages.edward</li>
    <li>python38Packages.flatbuffers</li>
    <li>python38Packages.gviz-api</li>
    <li>python38Packages.ignite</li>
    <li>python38Packages.mask-rcnn</li>
    <li>python38Packages.pyro-ppl</li>
    <li>python38Packages.pytorch (python38Packages.pytorchWithoutCuda)</li>
    <li>python38Packages.pytorch-lightning</li>
    <li>python38Packages.pytorch-metric-learning</li>
    <li>python38Packages.pytorchWithCuda</li>
    <li>python38Packages.pywick</li>
    <li>python38Packages.rising</li>
    <li>python38Packages.shap</li>
    <li>python38Packages.skorch</li>
    <li>python38Packages.slicer</li>
    <li>python38Packages.tensorboard-plugin-profile</li>
    <li>python38Packages.tensorboard-plugin-wit</li>
    <li>python38Packages.tensorboardx</li>
    <li>python38Packages.tensorflow (python38Packages.tensorflowWithoutCuda)</li>
    <li>python38Packages.tensorflow-estimator</li>
    <li>python38Packages.tensorflow-tensorboard</li>
    <li>python38Packages.test-tube</li>
    <li>python38Packages.tflearn</li>
    <li>python38Packages.torchvision</li>
    <li>python39Packages.baselines</li>
    <li>python39Packages.edward</li>
    <li>python39Packages.flatbuffers</li>
    <li>python39Packages.gviz-api</li>
    <li>python39Packages.ignite</li>
    <li>python39Packages.pyro-ppl</li>
    <li>python39Packages.pytorch (python39Packages.pytorchWithoutCuda)</li>
    <li>python39Packages.pytorch-lightning</li>
    <li>python39Packages.rising</li>
    <li>python39Packages.tensorboard-plugin-profile</li>
    <li>python39Packages.tensorboard-plugin-wit</li>
    <li>python39Packages.tensorflow (python39Packages.tensorflowWithoutCuda)</li>
    <li>python39Packages.tensorflow-estimator</li>
    <li>python39Packages.tensorflow-tensorboard</li>
    <li>python39Packages.tensorflowWithCuda</li>
    <li>python39Packages.tflearn</li>
    <li>python39Packages.torchvision</li>
  </ul>
</details>

Mic92 commented:
I could not get logs for some of the failures. Let me run it again.

ndl commented:
> I could not get logs for some of the failures. Let me run it again.

OK, so for python37 failures I think I know the reason: when tried locally, it failed on missing "typing-extensions" dependency during the build. "typing-extensions" seems to be built-in starting from python38 so I haven't noticed it during my testing, as all the previous tests I've done were with python38. I've added an explicit dependency to "typing-extensions" and updated the PR, please try it again.

The failures with python38, OTOH, are surprising as I've built locally both cuda and non-cuda versions successfully, so getting the error logs from these would be very useful indeed!

Mic92 commented:
Result of `nixpkgs-review pr 108923` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python37Packages.rl-coach</li>
    <li>python38Packages.rl-coach</li>
  </ul>
</details>
<details>
  <summary>27 packages failed to build:</summary>
  <ul>
    <li>python37Packages.arviz</li>
    <li>python37Packages.baselines</li>
    <li>python37Packages.dm-sonnet</li>
    <li>python37Packages.edward</li>
    <li>python37Packages.graph_nets</li>
    <li>python37Packages.mask-rcnn</li>
    <li>python37Packages.optuna</li>
    <li>python37Packages.pymc3</li>
    <li>python37Packages.tensorflow (python37Packages.tensorflow-build ,python37Packages.tensorflow-build_2 ,python37Packages.tensorflowWithoutCuda ,python37Packages.tensorflow_2)</li>
    <li>python37Packages.tensorflow-bin (python37Packages.tensorflow-bin_2)</li>
    <li>python37Packages.tensorflow-probability</li>
    <li>python37Packages.tensorflowWithCuda</li>
    <li>python37Packages.tflearn</li>
    <li>python37Packages.torchgpipe</li>
    <li>python38Packages.arviz</li>
    <li>python38Packages.dm-sonnet</li>
    <li>python38Packages.graph_nets</li>
    <li>python38Packages.optuna</li>
    <li>python38Packages.pymc3</li>
    <li>python38Packages.tensorflow-probability</li>
    <li>python38Packages.torchgpipe</li>
    <li>python39Packages.dm-sonnet</li>
    <li>python39Packages.graph_nets</li>
    <li>python39Packages.pytorch-metric-learning</li>
    <li>python39Packages.tensorflow-probability</li>
    <li>python39Packages.torchgpipe</li>
    <li>tts</li>
  </ul>
</details>
<details>
  <summary>64 packages built:</summary>
  <ul>
    <li>python37Packages.flatbuffers</li>
    <li>python37Packages.gviz-api</li>
    <li>python37Packages.ignite</li>
    <li>python37Packages.pyro-ppl</li>
    <li>python37Packages.pytorch (python37Packages.pytorchWithoutCuda)</li>
    <li>python37Packages.pytorch-lightning</li>
    <li>python37Packages.pytorch-metric-learning</li>
    <li>python37Packages.pytorchWithCuda</li>
    <li>python37Packages.pywick</li>
    <li>python37Packages.rising</li>
    <li>python37Packages.shap</li>
    <li>python37Packages.skorch</li>
    <li>python37Packages.slicer</li>
    <li>python37Packages.tensorboard-plugin-profile</li>
    <li>python37Packages.tensorboard-plugin-wit</li>
    <li>python37Packages.tensorboardx</li>
    <li>python37Packages.tensorflow-estimator (python37Packages.tensorflow-estimator_2)</li>
    <li>python37Packages.tensorflow-tensorboard (python37Packages.tensorflow-tensorboard_2)</li>
    <li>python37Packages.test-tube</li>
    <li>python37Packages.torchvision</li>
    <li>python38Packages.baselines</li>
    <li>python38Packages.edward</li>
    <li>python38Packages.flatbuffers</li>
    <li>python38Packages.gviz-api</li>
    <li>python38Packages.ignite</li>
    <li>python38Packages.mask-rcnn</li>
    <li>python38Packages.pyro-ppl</li>
    <li>python38Packages.pytorch (python38Packages.pytorchWithoutCuda)</li>
    <li>python38Packages.pytorch-lightning</li>
    <li>python38Packages.pytorch-metric-learning</li>
    <li>python38Packages.pytorchWithCuda</li>
    <li>python38Packages.pywick</li>
    <li>python38Packages.rising</li>
    <li>python38Packages.shap</li>
    <li>python38Packages.skorch</li>
    <li>python38Packages.slicer</li>
    <li>python38Packages.tensorboard-plugin-profile</li>
    <li>python38Packages.tensorboard-plugin-wit</li>
    <li>python38Packages.tensorboardx</li>
    <li>python38Packages.tensorflow (python38Packages.tensorflow-build ,python38Packages.tensorflow-build_2 ,python38Packages.tensorflowWithoutCuda ,python38Packages.tensorflow_2)</li>
    <li>python38Packages.tensorflow-estimator (python38Packages.tensorflow-estimator_2)</li>
    <li>python38Packages.tensorflow-tensorboard (python38Packages.tensorflow-tensorboard_2)</li>
    <li>python38Packages.tensorflowWithCuda</li>
    <li>python38Packages.test-tube</li>
    <li>python38Packages.tflearn</li>
    <li>python38Packages.torchvision</li>
    <li>python39Packages.baselines</li>
    <li>python39Packages.edward</li>
    <li>python39Packages.flatbuffers</li>
    <li>python39Packages.gviz-api</li>
    <li>python39Packages.ignite</li>
    <li>python39Packages.pyro-ppl</li>
    <li>python39Packages.pytorch (python39Packages.pytorchWithoutCuda)</li>
    <li>python39Packages.pytorch-lightning</li>
    <li>python39Packages.pytorchWithCuda</li>
    <li>python39Packages.rising</li>
    <li>python39Packages.tensorboard-plugin-profile</li>
    <li>python39Packages.tensorboard-plugin-wit</li>
    <li>python39Packages.tensorflow (python39Packages.tensorflow-build ,python39Packages.tensorflow-build_2 ,python39Packages.tensorflowWithoutCuda ,python39Packages.tensorflow_2)</li>
    <li>python39Packages.tensorflow-estimator (python39Packages.tensorflow-estimator_2)</li>
    <li>python39Packages.tensorflow-tensorboard (python39Packages.tensorflow-tensorboard_2)</li>
    <li>python39Packages.tensorflowWithCuda</li>
    <li>python39Packages.tflearn</li>
    <li>python39Packages.torchvision</li>
  </ul>
</details>

Mic92 commented:
> > I could not get logs for some of the failures. Let me run it again.
> 
> OK, so for python37 failures I think I know the reason: when tried locally, it failed on missing "typing-extensions" dependency during the build. "typing-extensions" seems to be built-in starting from python38 so I haven't noticed it during my testing, as all the previous tests I've done were with python38. I've added an explicit dependency to "typing-extensions" and updated the PR, please try it again.
> 
> The failures with python38, OTOH, are surprising as I've built locally both cuda and non-cuda versions successfully, so getting the error logs from these would be very useful indeed!

The latest run no longer breaks on python38. However python37 still breaks. I would say either add the typing-extensions or disable python3.7.

Mic92 commented:
This is the error log for python3.7-tensorflow: https://gist.github.com/Mic92/91fa5319eaf143d7da8ada7d6c21b928

ndl commented:
> The latest run no longer breaks on python38. However python37 still breaks. I would say either add the typing-extensions or disable python3.7.

Did you do the latest run on 69d1e62 commit? (= the last update to PR done yesterday?) It already includes "typing-extensions" fix and with this fix on my machine `python37Packages.tensorflow_2` builds fine both for CUDA and non-CUDA configurations.

> This is the error log for python3.7-tensorflow: https://gist.github.com/Mic92/91fa5319eaf143d7da8ada7d6c21b928

That error log looks quite strange: it references "tensorflow_cpu-2.1.0" whereas this PR is for tensorflow 2.4.0. Looking through Nixpkgs sources, it seems that the binary version of TensorFlow (the one available as `pythonPackages.tensorflow-bin`) uses version 2.1.0 indeed and is currently broken: it expects `gast=0.3.2` whereas the current version of gast in Nixpkgs master is `0.4.0`. Version 2.1.0 is also quite outdated (2020-01-08, that's now more than a year ago) and is very likely to be broken now due to multiple other dependencies versions mismatches (gast just happened to be the first one that triggered the error).

My best guess is that because this PR touches `tensorflow-estimator_2` and `tensorflow-tensorboard_2` (which are also the dependencies for `tensorflow-bin`), `tensorflow-bin` is pulled in for the build as well - but it was already broken before due to dependencies versions mismatches.

Shall I add in this PR `disabled` attribute to `tensorflow-bin` derivation with an explanation of the issue to cut `tensorflow-bin` from the build dependencies and so that any packages explicitly depending on `tensorflow-bin` (if there are any) can be detected as broken?

Mic92 commented:
> > The latest run no longer breaks on python38. However python37 still breaks. I would say either add the typing-extensions or disable python3.7.
> 
> Did you do the latest run on [69d1e62](https://github.com/NixOS/nixpkgs/commit/69d1e62b36bf1b65754acfd3a6f343b6b919dd25) commit? (= the last update to PR done yesterday?) It already includes "typing-extensions" fix and with this fix on my machine `python37Packages.tensorflow_2` builds fine both for CUDA and non-CUDA configurations.
> 
> > This is the error log for python3.7-tensorflow: [gist.github.com/Mic92/91fa5319eaf143d7da8ada7d6c21b928](https://gist.github.com/Mic92/91fa5319eaf143d7da8ada7d6c21b928)
> 
> That error log looks quite strange: it references "tensorflow_cpu-2.1.0" whereas this PR is for tensorflow 2.4.0. Looking through Nixpkgs sources, it seems that the binary version of TensorFlow (the one available as `pythonPackages.tensorflow-bin`) uses version 2.1.0 indeed and is currently broken: it expects `gast=0.3.2` whereas the current version of gast in Nixpkgs master is `0.4.0`. Version 2.1.0 is also quite outdated (2020-01-08, that's now more than a year ago) and is very likely to be broken now due to multiple other dependencies versions mismatches (gast just happened to be the first one that triggered the error).
> 
> My best guess is that because this PR touches `tensorflow-estimator_2` and `tensorflow-tensorboard_2` (which are also the dependencies for `tensorflow-bin`), `tensorflow-bin` is pulled in for the build as well - but it was already broken before due to dependencies versions mismatches.
> 
> Shall I add in this PR `disabled` attribute to `tensorflow-bin` derivation with an explanation of the issue to cut `tensorflow-bin` from the build dependencies and so that any packages explicitly depending on `tensorflow-bin` (if there are any) can be detected as broken?

I used nixpkgs-review, which merges your PR into the latest master.

ndl commented:
OK, so RE: the failure in the build log above - I've confirmed that when running `nixpkgs-review` on this PR it does pull in `tensorflow-bin` indeed and that's where this error comes from.

Also `tensorflow-bin` is indeed broken before this PR: running `nix-build -A python37Packages.tensorflow-bin` on Nixpkgs master produces exactly the same error.
